ai-in-directus-flows/index.md

```diff
@@ -0,0 +1,154 @@
+
+```json
+"usage": {
+    "completion_tokens": 17,
+    "prompt_tokens": 57,
+    "total_tokens": 74
+} 
+```
+
+Each prompt/response combination is stored in a Directus collection and various query filters and aggregators are used to provide the game engine with the total number of tokens used by a player in a day and the total tokens used for the current game.
+
+If the player reaches their daily limit, no more Chat GPT prompts can be sent to the API and the game will prompt you to wait until tomorrow before being able to progress.
+
+Once the total game token limit is reached the prompt sent to Chat GPT is modified such that the next responses will include a satisfying conclusion to the game currently in progress.
+
+## The game engine
+The [AIVenture][14] game engine is responsible for:
+
+1. Providing initial context to Chat GPT (eg. “You are a text adventure game writer…”).
+2. Generating an introduction (eg. “You are standing in an open field…”).
+3. Responding to player prompts in a way compliant with any game rules and the initial context.
+4. Managing Chat GPT interactions including token counts and unexpected responses.
+5. Providing a satisfactory end game state.
+6. State management for games in progress.
+
+In AIVenture all this logic is represented in Directus Flows. However with the large number of operations required we need to plan our flow architecture carefully to avoid unnecessarily complicating our game engine flow.
+
+### Managing complex flows.
+
+Creating a game engine in Directus Flows has the potential to become excessively large and unmaintainable very quickly without some strict architectural decisions. 
+
+We keep flows as small and functionally isolated components. Think of these flows as you would Classes in your application. We then have a central or main flow that triggers our sub-flows. This is a powerful and important architecture for our game engine design as it allows us to compartmentalise logic, keeping our main flow clean and readable and making testing easier. 
+
+Let’s see what this looks like in Directus. First, we set up a flow called “SubCommand”. Our one important configuration for this flow is to select “Another flow” as the trigger set up - this will allow us to trigger this sub flow from within our parent and receive the resulting data. All our flow logic for our subcommand can now be added, being sure to return the required data.
+
+![Selecting “Another Flow” as our Trigger][image-1]
+
+Next we want to add our master or parent flow. This will be called from our frontend so should have a “Webhook” trigger - for ease of testing in the browser, let’s use the GET method. Then we add a new operation to our flow of type “Trigger Flow” and select our “SubCommand” flow from above. Our payload should include the data from the subcommand, ie. `{{ $last }}`.
```

Comment:
```suggestion
Next, we want to add our central flow. This will be called from the frontend so should have a “Webhook” trigger. For ease of testing in the browser, use the GET method. Then, we add a new operation to our flow of type “Trigger Flow” and select our “SubCommand” flow from above. Our payload should include the data from the subcommand, ie. `{{ $last }}`.
```

Standardizing earlier language, but also `master` is a term we generally keep our of projects and writing where possible.

Reply:
Understood. Perhaps "main" instead of "central" as this flow is really the top most piece of logic. Thoughts?

Reply:
Yup works for me. It was more the rejection of the word master than a promotion of central. Whatever works for you.